src/details/ArborX_Ray.hpp

```diff
@@ -136,6 +148,96 @@ bool intersects(Ray const &ray, Box const &box)
   return max_min <= min_max && (min_max >= 0);
 }
 
-} // namespace Details
+KOKKOS_INLINE_FUNCTION float dotProduct(Ray::Vector const &v1,
+                                        Ray::Vector const &v2)
+{
+  return v1[0] * v2[0] + v1[1] * v2[1] + v1[2] * v2[2];
+}
+
+// Solves a*x^2 + b*x + c = 0.
+// If a solution exists, return true and stores roots at x1, x2.
+// If a solution does not exist, returns false.
+KOKKOS_INLINE_FUNCTION bool solveQuadratic(float const a, float const b,
+                                           float const c, float &x1, float &x2)
+{
+  assert(a != 0);
+
+  auto const discriminant = b * b - 4 * a * c;
+  if (discriminant < 0)
+    return false;
+  if (discriminant == 0)
+  {
+    x1 = x2 = -b / (2 * a);
+    return true;
+  }
+
+  // Instead of doing a simple
+  //    (-b +- std::sqrt(discriminant)) / (2*a)
+  // we use a more stable algorithm with less loss of precision (see, for
+  // example, https://www.scratchapixel.com/lessons/3d-basic-rendering/
+  // minimal-ray-tracer-rendering-simple-shapes/ray-sphere-intersection).
+  auto const sgn = [](float x) { return (int)(x > 0) - (int)(x < 0); };
+  auto const q = 0.5 * (-b - sgn(b) * std::sqrt(discriminant));
+  x1 = q / a;
+  x2 = c / q;
```

Comment:
@dalg24 discovered that this code contains a bug. Specifically, when `b = 0`, `q = inf`, and the roots are wrong. This case has to be treated specially.

Reply:
b=0 alone is not a problem, maybe sgn(b) returning 0 is an issue. We can use the the below implementation instead:
 q = (b > 0) ? 
            -0.5 * (b + sqrt(discriminant)) : 
            -0.5 * (b - sqrt(discriminant));

Reply:
Yeah, I think this could work, too. You had it originally, and I tried to be cleaner and introduced a bug.

Reply:
I would change it to
```c++
  q = (b > 0) ?  (-b - sqrt(discriminant))/2 : (-b + sqrt(discriminant))/2
```
to make it more in line with the standard formula `(-b +- sqrt(discriminant))/(2*a)`